File: GFG POTD/022-September-2025/028-longest-bounded-difference-subarray.cpp
// https://www.geeksforgeeks.org/problems/longest-bounded-difference-subarray/1/




#include <bits/stdc++.h>
using namespace std;


class Solution {
  private:
    using P = pair<int, int>;
    
    // TC : O(N*log(N))
    // SC : O(N)
    vector<int> approach_1(vector<int>& arr, int x) {
        int n = arr.size();
        priority_queue<P, vector<P>, greater<P>> mnH;
        priority_queue<P> mxH;
        int maxLen = 0;
        int start  = 0;
        int i = 0, j = 0;
        while (j < n) {
            mnH.push({arr[j], j});
            mxH.push({arr[j], j});
            while (abs(mnH.top().first - mxH.top().first) > x) {
                while (!mnH.empty() && mnH.top().second <= i) mnH.pop();
                while (!mxH.empty() && mxH.top().second <= i) mxH.pop();
                i ++;
            }
            if (j - i + 1 > maxLen) {
                maxLen = (j - i + 1);
                start  = i;
            }
            j ++;
        }
        vector<int> result;
        for (int i = start; i < maxLen + start; i ++) {
            result.push_back(arr[i]);
        }
        return result;
    }
    
    // TC : O(N)
    // SC : O(N)
    vector<int> approach_2(vector<int>& arr, int x) {
        int n = arr.size();
        deque<int> mnDq, mxDq;
        int maxLen = 0;
        int start  = 0;
        int i = 0, j = 0;
        while (j < n) {
            while (!mnDq.empty() && arr[j] <= arr[mnDq.back()]) { // curr element is smaller than minQ back
                mnDq.pop_back();
            }
            while (!mxDq.empty() && arr[j] >= arr[mxDq.back()]) { // curr element is greater than maxQ back
                mxDq.pop_back();
            }
            
            mnDq.push_back(j);
            mxDq.push_back(j);
            
            while (abs(arr[mxDq.front()] - arr[mnDq.front()]) > x) {
                if (mnDq.front() == i) mnDq.pop_front();    // catch
                if (mxDq.front() == i) mxDq.pop_front();
                i ++;
            }
            if (j - i + 1 > maxLen) {
                maxLen = (j - i + 1);
                start  = i;
            }
            j ++;
        }
        vector<int> result;
        for (int i = start; i < maxLen + start; i ++) {
            result.push_back(arr[i]);
        }
        return result;
    }
  public:
    vector<int> longestSubarray(vector<int>& arr, int x) {
        // return approach_1(arr, x);
        return approach_2(arr, x);
    }
};

File: GFG POTD/022-September-2025/028-longest-bounded-difference-subarray.java
// https://www.geeksforgeeks.org/problems/longest-bounded-difference-subarray/1/




import java.util.*;

class Solution {
    // TC : O(N*log(N))
    // SC : O(N)
    private ArrayList<Integer> approach_1(int[] arr, int x) {
        int n = arr.length;
        PriorityQueue<Integer> minH = new PriorityQueue<>((a, b) -> arr[a] - arr[b]);
        PriorityQueue<Integer> maxH = new PriorityQueue<>((a, b) -> arr[b] - arr[a]);
        
        int maxLen   = 0;
        int startIdx = 0;
        int i = 0, j = 0;
        
        while (j < n) {
            minH.add(j);
            maxH.add(j);
            
            while (Math.abs(arr[minH.peek()] - arr[maxH.peek()]) > x) {
                while (!minH.isEmpty() && minH.peek() <= i) minH.poll();
                while (!maxH.isEmpty() && maxH.peek() <= i) maxH.poll();
                i ++;
            }
            
            if (j - i + 1 > maxLen) {
                maxLen = (j - i + 1);
                startIdx = i;
            }
            j ++;
        }
        
        ArrayList<Integer> result = new ArrayList<>();
        for (i = startIdx; i < startIdx + maxLen; i ++) {
            result.add(arr[i]);
        }
        return result;
    }
    
    // TC : O(N)
    // SC : O(N)
    public ArrayList<Integer> approach_2(int[] arr, int x) {
        int n = arr.length;
        Deque<Integer> minD = new ArrayDeque<>();
        Deque<Integer> maxD = new ArrayDeque<>();
        
        int maxLen   = 0;
        int startIdx = 0;
        int i = 0, j = 0;
        
        while (j < n) {
            while (!minD.isEmpty() && arr[j] <= arr[minD.peekLast()]) minD.pollLast();
            while (!maxD.isEmpty() && arr[j] >= arr[maxD.peekLast()]) maxD.pollLast();
            
            minD.add(j);
            maxD.add(j);
            
            while (Math.abs(arr[minD.peekFirst()] - arr[maxD.peekFirst()]) > x) {
                if (minD.peekFirst() == i) minD.pollFirst();
                if (maxD.peekFirst() == i) maxD.pollFirst();
                i ++;
            }
            
            if (j - i + 1 > maxLen) {
                maxLen = (j - i + 1);
                startIdx = i;
            }
            j ++;
        }
        
        ArrayList<Integer> result = new ArrayList<>();
        for (i = startIdx; i < startIdx + maxLen; i ++) {
            result.add(arr[i]);
        }
        return result;
    }
    public ArrayList<Integer> longestSubarray(int[] arr, int x) {
        // return approach_1(arr, x);
        return approach_2(arr, x);
    }
}